torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,399 @@
+    node->addInput(alloc->output());
+    alloc->insertBefore(node);
+    alloc->addInput(slab->output());
+
+    auto ttp = val->type()->expect<c10::TensorType>();
+    alloc->ty_(attr::profiled_type, ttp);
+    alloc->i_(attr::size, (int64_t)region.size);
+    alloc->i_(attr::offset, (int64_t)region.offset);
+    release_slab->addInput(alloc->output());
+  }
+  release_slab->insertBefore(graph->return_node());
+}
+
+bool hasOutVariant(Node* node) {
+  if (!node->maybeSchema()) {
+    return false;
+  }
+  const auto& node_schema_args = node->schema().arguments();
+  for (const auto& variant : getAllOperatorsFor(node->kind())) {
+    auto variant_schema = variant->schema();
+    // a simple check for an out param is insufficient because ops with multiple
+    // out variants e.g. aten::cat.out and aten::cat.names_out
+    auto maybe_out_arg_idx = variant_schema.argumentIndexWithName("out");
+    if (maybe_out_arg_idx) {
+      // out arg should be the *next* arg after current args
+      if ((size_t)maybe_out_arg_idx.value() != node_schema_args.size()) {
+        return false;
+      }
+      // functions are contravariant in arguments i.e. args should be "broader"
+      // in replacement fn i.e. to check variant_schema.args[] <:
+      // node_schema.args[] we check that node_schema.args[i] <:
+      // variant_schema[i] for all i <= len(node_schema.args)
+      c10::isSubtypeOfList(
+          ArrayRef<Argument>(node_schema_args),
+          ArrayRef<Argument>(variant_schema.arguments())
+              // e.g. maybe_out_arg_idx = 3 means first 3 arg types should
+              // satisfy subtype relationship
+              .slice(0, maybe_out_arg_idx.value()),
+          nullptr);
+      return true;
+    }
+  }
+  return false;
+}
+
+std::pair<
+    FastMap<const Value*, std::pair<UniqueLiveRange, size_t>>,
+    FastMap<const Value*, std::pair<UniqueLiveRange, size_t>>>
+getManagedAndUnManagedValues(
+    const std::shared_ptr<Graph>& graph,
+    bool frozen = false) {
+  AliasDb alias_db(graph, frozen);
+  FastSet<const Value*> always_alive_values =
+      jit::GetAlwaysAliveValues(graph, alias_db);
+  FastMap<const Value*, LiveRange> live_ranges =
+      jit::GetLiveness(graph, always_alive_values, alias_db).second;
+
+  FastMap<const Value*, std::pair<UniqueLiveRange, size_t>> managed_values;
+  FastSet<const Value*> leaked_values;
+  FastMap<Node*, bool> node_has_out_variant;
+  for (auto* node : graph->nodes()) {
+    auto has_out = hasOutVariant(node);
+    node_has_out_variant.insert({node, has_out});
+  }
+
+  for (auto node : graph->nodes()) {
+    if (!node_has_out_variant[node]) {
+      leaked_values.insert(node->outputs().begin(), node->outputs().end());
+      continue;
+    }
+    for (const auto* out_v : node->outputs()) {
+      if (always_alive_values.count(out_v)) {
+        continue;
+      }
+      auto size = computeStorageSize(*out_v);
+      if (size > 0 && !isOptimizableContainerType(node, node_has_out_variant)) {
```

Comment:
What does isOptimizableContainerType mean here ? Name is a little vague .. maybe add comment